File: data/derived_cot/rq1_traces/translation/r1/cpp_to_py/code_output/DiscountStrategy.py
from enum import Enum

class DiscountStrategy:
    class PromoType(Enum):
        FidelityPromo = 1
        BulkItemPromo = 2
        LargeOrderPromo = 3
        NoPromo = 4

    def __init__(self, customer, cart, promo=PromoType.NoPromo):
        self.customer_ = customer
        self.cart_ = cart
        self.promo_ = promo

    def total(self):
        total_val = 0.0
        for item in self.cart_:
            total_val += item["quantity"] * item["price"]
        return total_val

    def due(self):
        return self.total() - self.promotion(self)

    def promotion(self, order):
        if self.promo_ == DiscountStrategy.PromoType.FidelityPromo:
            return self._fidelity_discount()
        elif self.promo_ == DiscountStrategy.PromoType.BulkItemPromo:
            return self._bulk_item_discount()
        elif self.promo_ == DiscountStrategy.PromoType.LargeOrderPromo:
            return self._large_order_discount()
        else:
            return 0.0

    def _fidelity_discount(self):
        fidelity = self.customer_["fidelity"]
        if fidelity >= 1000.0:
            return 0.05 * self.total()
        return 0.0

    def _bulk_item_discount(self):
        discount = 0.0
        for item in self.cart_:
            if item["quantity"] >= 20:
                discount += item["quantity"] * item["price"] * 0.1
        return discount

    def _large_order_discount(self):
        num_items = len(self.cart_)
        if num_items >= 10:
            return 0.07 * self.total()
        return 0.0

File: data/derived_cot/rq2_eval/translation/r1/_converted_code/py_to_cpp/DiscountStrategy.cpp
#include <vector>
#include <set>
#include <string>

class DiscountStrategy {
public:
    struct Customer {
        std::string name;
        int fidelity;
    };

    struct Item {
        std::string product;
        int quantity;
        double price;
    };

    using Promotion = double (*)(DiscountStrategy*);

    DiscountStrategy(const Customer& customer, const std::vector<Item>& cart, Promotion promotion = nullptr)
        : customer_(customer), cart_(cart), promotion_(promotion) {
        total_ = total();
    }

    double total() {
        total_ = 0.0;
        for (const auto& item : cart_) {
            total_ += item.quantity * item.price;
        }
        return total_;
    }

    double due() {
        double discount = 0.0;
        if (promotion_) {
            discount = promotion_(this);
        }
        return total_ - discount;
    }

    static double fidelity_discount(DiscountStrategy* order) {
        if (order->customer_.fidelity >= 1000) {
            return order->total() * 0.05;
        }
        return 0.0;
    }

    static double bulk_item_discount(DiscountStrategy* order) {
        double discount = 0.0;
        for (const auto& item : order->cart_) {
            if (item.quantity >= 20) {
                discount += item.quantity * item.price * 0.1;
            }
        }
        return discount;
    }

    static double large_order_discount(DiscountStrategy* order) {
        std::set<std::string> distinct;
        for (const auto& item : order->cart_) {
            distinct.insert(item.product);
        }
        if (distinct.size() >= 10) {
            return order->total() * 0.07;
        }
        return 0.0;
    }

private:
    Customer customer_;
    std::vector<Item> cart_;
    Promotion promotion_;
    double total_;
};